File: core/src/main/scala/io/eels/HdfsIterator.scala
package io.eels

import org.apache.hadoop.fs.{Path, LocatedFileStatus, RemoteIterator}

object HdfsIterator {
  def apply(iterator: RemoteIterator[LocatedFileStatus]): Iterator[LocatedFileStatus] = new Iterator[LocatedFileStatus] {
    override def hasNext: Boolean = iterator.hasNext
    override def next(): LocatedFileStatus = iterator.next()
  }
}

object PathIterator {
  def apply(path: Path): Iterator[Path] = {
    Iterator.iterate(path)(_.getParent).takeWhile(_ != null).toList.reverseIterator
  }
}

File: eel-core/src/main/kotlin/io/eels/HdfsIterator.kt
package io.eels

import org.apache.hadoop.fs.LocatedFileStatus
import org.apache.hadoop.fs.RemoteIterator

object HdfsIterator {
  operator fun invoke(iterator: RemoteIterator<LocatedFileStatus>): Iterator<LocatedFileStatus> = object : Iterator<LocatedFileStatus> {
    override fun hasNext(): Boolean = iterator.hasNext()
    override fun next(): LocatedFileStatus = iterator.next()
  }
}